// @tmagic/cli@1.3.0-beta.2/lib/Core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const utils_1 = require("./utils");
class Core {
    version = require('../package.json').version;
    options;
    moduleMainFilePath = {
        componentMap: {},
        pluginMap: {},
        configMap: {},
        valueMap: {},
        eventMap: {},
        datasourceMap: {},
        dsConfigMap: {},
        dsValueMap: {},
        dsEventMap: {},
    };
    dir = {
        temp: () => path_1.default.resolve(this.options.source, this.options.temp),
    };
    constructor(options) {
        this.options = options;
    }
    async writeTemp(file, content) {
        await fs_extra_1.default.outputFile(path_1.default.resolve(this.dir.temp(), file), content);
    }
    async init() {
        this.moduleMainFilePath = (0, utils_1.resolveAppPackages)(this);
        if (typeof this.options.onInit === 'function') {
            this.moduleMainFilePath = await this.options.onInit(this);
        }
    }
    async prepare() {
        await (0, utils_1.prepareEntryFile)(this);
        if (typeof this.options.onPrepare === 'function') {
            this.options.onPrepare(this);
        }
    }
}
exports.default = Core;
